text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,590 @@
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
+the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
+"Receive" span.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
+MAY link to the "Create" or "Publish" span for the message.
+
+## Examples
+
+This section contains a list of examples illustrating the use of the
+conventions outlined above. Green boxes denote spans that are required to
+exist in order to conform to those conventions. Other boxes denote spans that
+are not required and covered by the conventions, but are hopefully helpful in
+understanding how messaging spans can be integrated into an overall trace flow.
+
+### Single message producer, single message push-based consumer
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0 color:green,stroke:green
+```
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+  PM1-->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+### Single message producer, single message push-based consumer with manual settlement
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+    direction TB
+    PM1[Publish m1]
+  end
+  subgraph CONSUMER
+    direction TB
+    DM1[Deliver m1]-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1,S1 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+### Single message producer, single message push-based consumer with auto-settlement
```

Comment:
Since these are the same, maybe we could combine both `###` sections under the same diagram?

Reply:
Done.